kubernetes/updater/pkg/maintenance/trigger.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package maintenance
+
+import (
+	"context"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Trigger is evaluated to decide whether a maintenance can happen or not.
+// Maintenances can happen because of multiple reasons like:
+// - attempt to recover from a broken state
+// - we are in a maintenance window
+// - emergency security patch
+// Each Trigger has a Name() for logging purposes and a Default() method
+// returning whether the trigger should fail open or close in case of error.
```

Comment:
Sorry, it was not clear to me what it means to "fail open or close in case of error"

Does it mean that when calling `AtLeastOne`, when the `CanStart` fails, it will override the value with the one provided by `Default()`?

Reply:
Yeah, if we can't evaluate CanStart() because of some reason, what do we want to default to? For example if there's no upgrade schedule exported by the agent we want to trigger a maintenance because this means the agent is unhealthy. But if we can't real our s3 bucket because of some reasons we don't want to start a maintenance.